### PR TITLE
[wip]: android - keep screen awake toggle

### LINKED ISSUE #2748: Android: Keep lights on
First, congratulations for releasing an Android version.

However, on Android it usually doesn't make sense to have a menu item that says E-ink optimization. It makes much better sense to have settings to adjust backlight and to block it from dimming. Then again, some e-ink devices have indeed Android.

### LINKED ISSUE #4364: Keep screen awake still not available
* KOReader version: 2018.11.1
* Device: Galaxy S7

#### Issue

I commented here https://github.com/koreader/koreader/issues/2748 that an option to keep a device screen from turning off is still not available despite a commit to to fix.

### PR DESCRIPTION
**Pull request by pazos**
Fixes #2748 
Fixes #4364 

It works but, when enabled for the first time, it turns the screen black until an input event is registered. Further toggles don't show this issue.

Maybe @Frenzie knows ;)

### CODE REVIEW

**Comment by Frenzie** [comment]
@pazos There's a long discussion about something similar in #3118.

We're probably doing something wrong with ANativeWindow but it's not clear to me what. It's also difficult for me to test, because:

1. This doesn't occur in the emulators, which afaik are the only option to develop for Android that even remotely approaches anything vaguely reminiscent of sanity.
2. The top half of the touch screen on my old old Android phone (Xperia Ray) that was the only way to test some of these issues somehow broke in recent months from just sitting around, plus it's very slow.
3. My 2015 Android phone (Zenfone 2) doesn't run KOReader, neither the x86 nor the ARM version. Since debugging on Android is basically impossible without root and I don't want to break my main phone that I use for, you know, actually important life things I haven't been able to figure out what's wrong there.
4. I already nuked Android on the Xperia XA2 I just bought and put Sailfish on it. It's actually amazing how much nicer it is than Android 8 (let alone Android <8), and it's mind-boggling that they don't have a paid app store. I don't understand why Nokia dropped this, or why Microsoft dropped their also superior mobile OS. I used Windows Mobile 6 on an HTC Touch Diamond 2 recently and it's still so much better than Android.

So what do I have Android phones for? Well, the hardware improvements still trump the inferior OS. /rant

**Comment by Frenzie** [comment]
> you seem a reaaaally early adopter since the image for the SonyXA2 isn't available on the Jolla store right now.

But it is? Note the top left, not the bottom left. ;-)

998df678-eae2-4a92-ae48-d03866516797

> BTW, It seems that some logging from android-luajit-launcher/blob/master/assets/android.lua is wrong, like the line in the link.

Wrong how?

**Comment by pazos** [comment]
> Wrong how?

Trying to print boolean values

```diff
diff --git a/assets/android.lua b/assets/android.lua
index 33ec818..820cd73 100644
--- a/assets/android.lua
+++ b/assets/android.lua
@@ -1191,6 +1191,8 @@ end
 function android.get_application_directory()
 end
 
+function boolName(value) return value and "true" or "false" end
+
 --[[
 the C code will call this function:
 --]]
@@ -1370,7 +1372,7 @@ local function run(android_app_state)
         end)
     end
     android.setFullscreen = function(fullscreen)
-        android.LOGI("setting fullscreen ", fullscreen)
+        android.LOGI("Switching fullscreen to " .. boolName(fullscreen))
         JNI:context(android.app.activity.vm, function(JNI)
             JNI:callVoidMethod(
                 android.app.activity.clazz,
@@ -1401,7 +1403,6 @@ local function run(android_app_state)
                 "hasClipboardTextIntResultWrapper",
                 "()I"
             )
-            android.LOGI("Has text in clipboard: ", hasClipboardText)
             return hasClipboardText == 1
         end)
     end
@@ -1421,7 +1422,7 @@ local function run(android_app_state)
     end
 
     android.setKeepScreenOn = function(keepOn)
-        android.LOGI("setting KeepScreenOn to ", keepOn)
+        android.LOGI("Switching screen timeout to " .. boolName(not keepOn))
         JNI:context(android.app.activity.vm, function(JNI)
             JNI:callVoidMethod(
                 android.app.activity.clazz,
@@ -1439,13 +1440,12 @@ local function run(android_app_state)
                 "isWifiEnabled",
                 "()I"
             )
-            android.LOGI("is WifiEnabled =", isWifiEnabled)
             return isWifiEnabled == 1
         end)
     end
 
     android.setWifiEnabled = function(wifiEnabled)
-        android.LOGI("setting wifi to: ", wifiEnabled)
+        android.LOGI("Switching WiFi to " ..  boolName(wifiEnabled))
         JNI:context(android.app.activity.vm, function(JNI)
             JNI:callVoidMethod(
                 android.app.activity.clazz,
```

**Comment by Frenzie** [comment]
> Trying to print boolean values

And what happens? What does the output look like? 

Are you being overly complex when a simple `tostring()` would be better? ;-)

**Comment by Frenzie** [comment]
So it's being set to 1. Looks like intended to me? I wouldn't obfuscate what's really happening on the Android end tbh, plus your example function conflates nil and false. :-)

I'd say just make android.LOGI take multiple arguments, see:

https://github.com/koreader/koreader/blob/c41dfc25dd40a40c239a0a3289adaf0989849a29/frontend/logger.lua#L43-L58

**Comment by pazos** [comment]
@Frenzie: yup, I got it wrong, tostring(bool) will print true or false. My bad.

```lua
#!/usr/bin/env luajit

a=1
b="something"
c={}
d=true
e=nil

print(tostring(a) .. " " .. type(a))
print(tostring(b) .. " " .. type(b))
print(tostring(c) .. " " .. type(c))
print(tostring(d) .. " " .. type(d))
print(tostring(e) .. " " .. type(e))
```

I can't replicate the black screen issue on aosp 8.0/8.1. The switch works without issues.

**Comment by Frenzie** [comment]
On what version did it occur for you?

**Comment by pazos** [comment]
moto g5s - android 7.1.1

File: frontend/ui/elements/screen_keep_on_menu_table.lua [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+if not isAndroid then return end
+
+return {
+    text = _("Keep screen on"),
+    checked_func = function()
+        return G_reader_settings:isTrue("keep_screen_on")
+    end,
+    callback = function()
+        local keep_screen_on = G_reader_settings:isTrue("keep_screen_on")
+        android.setKeepScreenOn(not keep_screen_on)
```
Comment (Frenzie) [inline frontend/ui/elements/screen_keep_on_menu_table.lua]: Have you tried just doing it a couple of times here? E.g., the same line twice or 

```
android.setKeepScreenOn(not keep_screen_on)
android.setKeepScreenOn(keep_screen_on)
android.setKeepScreenOn(not keep_screen_on)
```

Sure, it'd be ugly, but if it works around the problem worse things have happened. ;-)

Comment (pazos) [reply frontend/ui/elements/screen_keep_on_menu_table.lua]: It seems a bad idea, it breaks the app on both devices. :astonished:

**Comment by pazos** [comment]
The broken behaviour is announced on logcat

```adb
12-26 21:17:01.050 23959 23972 I KOReader: DEBUG input event => type: 4, code: 2(nil), value: 1, time: 1545855421.50036
12-26 21:17:01.050 23959 23972 I KOReader: DEBUG Android application event 2
12-26 21:17:01.050 23959 23972 I KOReader: DEBUG input event => type: 4, code: 1(nil), value: 1, time: 1545855421.50396
12-26 21:17:01.050 23959 23972 I KOReader: DEBUG Android application event 1
```
 https://github.com/koreader/android-luajit-launcher/blob/master/assets/android.lua#L888-L889.

**Comment by pazos** [comment]
~~Lurking for other apps' behaviour I found that, for instance, Adobe Acrobat Reader just set the flag at the beginning of the activity and didn't offer a related setting.~~ It has a "block brightness" setting which does that"

 If we want to do that it is easier (at least for me) than trying to figure out a switch that works on most devices.

On the other side this PR, as is, should work on most devices without terminating the current window.

**Pull request closed without merging**